### PR TITLE
Include project-id in sync'ed data to allow future calls to directly jump to that project.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs [hunk 1]
```diff
@@ -156,10 +157,22 @@ public async Task<Checksum> GetChecksumAsync(ProjectId projectId, CancellationTo
...
+                        {
+                            projectIds.Add(projectStateChecksums.ProjectId);
+                            projectChecksums.Add(projectStateChecksums.Checksum);
+                        }
+                    }
+
                     return new SolutionStateChecksums(
                         attributesChecksum,
-                        new ChecksumCollection(projectChecksums.WhereNotNull().ToImmutableArray()),
+                        projectIds.ToImmutableAndClear(),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:174]: passing along the project ids, not just their checksums.  that way when we jhave an instance of SolutionStateChecksums we can scope requests down to a particular id when trying to retrieve a particular project checksum.

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs [hunk 2]
```diff
@@ -14,27 +15,43 @@
...
-    Checksum frozenSourceGeneratedDocumentIdentity,
-    Checksum frozenSourceGeneratedDocumentText) : IChecksummedObject
+internal sealed class SolutionStateChecksums : IChecksummedObject
 {
-    public Checksum Checksum { get; } = Checksum.Create(stackalloc[]
+    public Checksum Checksum { get; }
+
+    public Checksum Attributes { get; }
+    public ImmutableArray<ProjectId> ProjectIds { get; }
+    public ChecksumCollection Projects { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:24]: note: i merge ProejctIds and Projects into a singular collection type in a later PR to keep these more tightly bound

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs [hunk 3]
```diff
@@ -27,20 +28,20 @@ public async Task<SolutionInfo> CreateSolutionInfoAsync(Checksum solutionChecksu
         var solutionAttributes = await GetAssetAsync<SolutionInfo.SolutionAttributes>(hintProject: null, solutionChecksums.Attributes, cancellationToken).ConfigureAwait(false);
 
         using var _ = ArrayBuilder<ProjectInfo>.GetInstance(solutionChecksums.Projects.Count, out var projects);
-        foreach (var projectChecksum in solutionChecksums.Projects)
-            projects.AddIfNotNull(await CreateProjectInfoAsync(projectChecksum, cancellationToken).ConfigureAwait(false));
+        for (int i = 0, n = solutionChecksums.ProjectIds.Length; i < n; i++)
+            projects.AddIfNotNull(await CreateProjectInfoAsync(solutionChecksums.ProjectIds[i], solutionChecksums.Projects[i], cancellationToken).ConfigureAwait(false));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/AbstractAssetProvider.cs:32]: the benefit comes here.

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs [hunk 4]
```diff
@@ -27,20 +28,20 @@ public async Task<SolutionInfo> CreateSolutionInfoAsync(Checksum solutionChecksu
...
         return SolutionInfo.Create(
             solutionAttributes.Id, solutionAttributes.Version, solutionAttributes.FilePath, projects.ToImmutableAndClear(), analyzerReferences).WithTelemetryId(solutionAttributes.TelemetryId);
     }
 
-    public async Task<ProjectInfo?> CreateProjectInfoAsync(Checksum projectChecksum, CancellationToken cancellationToken)
+    public async Task<ProjectInfo?> CreateProjectInfoAsync(ProjectId projectId, Checksum projectChecksum, CancellationToken cancellationToken)
     {
-        var projectChecksums = await GetAssetAsync<ProjectStateChecksums>(hintProject: null, projectChecksum, cancellationToken).ConfigureAwait(false);
+        var projectChecksums = await GetAssetAsync<ProjectStateChecksums>(hintProject: projectId, projectChecksum, cancellationToken).ConfigureAwait(false);
+        Contract.ThrowIfFalse(projectId == projectChecksums.ProjectId);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/AbstractAssetProvider.cs:43]: now, when asking to retrieve the ProjectStateChecksums we don't need to do a linear walk to find it.  we can directly grab it using the known projectId.

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 5]
```diff
@@ -71,7 +71,8 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
 
                     if (oldSolutionChecksums.Projects.Checksum != newSolutionChecksums.Projects.Checksum)
                     {
-                        solution = await UpdateProjectsAsync(solution, oldSolutionChecksums.Projects, newSolutionChecksums.Projects, cancellationToken).ConfigureAwait(false);
+                        solution = await UpdateProjectsAsync(
+                            solution, oldSolutionChecksums, newSolutionChecksums, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:75]: i preferred this form as just passing around CheckasumCollection instances is unclear as to what they represent.

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs [hunk 6]
```diff
@@ -61,9 +79,10 @@ public static SolutionStateChecksums Deserialize(ObjectReader reader)
     {
         var checksum = Checksum.ReadFrom(reader);
         var result = new SolutionStateChecksums(
-            attributesChecksum: Checksum.ReadFrom(reader),
-            projectChecksums: ChecksumCollection.ReadFrom(reader),
-            analyzerReferenceChecksums: ChecksumCollection.ReadFrom(reader),
+            attributes: Checksum.ReadFrom(reader),
+            reader.ReadArray(ProjectId.ReadFrom),
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:83]: >reader.ReadArray(ProjectId.ReadFrom), [](http://example.com/codeflow?start=12&length=37)

nit: named param to match others would be nice

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs [hunk 7]
```diff
@@ -102,39 +121,48 @@ public static SolutionStateChecksums Deserialize(ObjectReader reader)
...
-
-            if (hintProject != null && hintProject != projectId)
-                continue;
-
-            if (projectState.TryGetStateChecksums(out var projectStateChecksums) &&
-                searchingChecksumsLeft.Remove(projectStateChecksums.Checksum))
+            var projectState = state.GetProjectState(hintProject);
+            if (projectState != null &&
+                projectState.TryGetStateChecksums(out var projectStateChecksums))
             {
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:132]: a searchingChecksumsLeft.Remove call isn't required here?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs [hunk 8]
```diff
@@ -156,10 +157,22 @@ public async Task<Checksum> GetChecksumAsync(ProjectId projectId, CancellationTo
                     var analyzerReferenceChecksums = ChecksumCache.GetOrCreate<ChecksumCollection>(AnalyzerReferences,
                         _ => new ChecksumCollection(AnalyzerReferences.SelectAsArray(r => serializer.CreateChecksum(r, cancellationToken))));
 
-                    var projectChecksums = await Task.WhenAll(projectChecksumTasks).ConfigureAwait(false);
+                    var allResults = await Task.WhenAll(projectChecksumTasks).ConfigureAwait(false);
+                    using var _1 = ArrayBuilder<ProjectId>.GetInstance(projectChecksumTasks.Length, out var projectIds);
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:161]: >projectChecksumTasks [](http://example.com/codeflow?start=71&length=20)

tiny nit: iterate over allResults below, so might be more clear to allocate based on it's Length

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:161]: will do!

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs]: done!

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:132]: no.  this isn't removing.  this is just fast-pathing finding the project-state-checksums to dive into it to search the actual checksums.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**